DIPs/DIP1xxx-rn.md

```diff
@@ -0,0 +1,927 @@
+in terms of compilation time, since no code is generated at all, therefore it is the one
+chosen in the design of the copy constructor.
+
+#### Overload resolution in the presence of parameters with copy constructors
+
+For a function in an overload set each argument is matched against the parameter the following way:
+
+1. Check if the argument type is a direct match to the parameter type.
+2. If not, check if the argument type is implicitly convertible to the parameter type.
+3. If not, check if the argument has a copy constructor that converts it to the parameter type.
+
+In the first 2 situations, if the check is true for a specific overload and that overload
+is selected to be called, if any of the parameters have a matching copy constructor, it will get called,
+otherwise, standard copying is employed. In the third situation, the overload is matched based on the
+conversion resulted from the copy constructor, therefore it the copy constructor will surely get called:
+
+```d
+struct A
+{
+    int[] a;
+    @implicit this(ref shared A) immutable {}
+}
+
+void fun(const A a) {}                  // 1
+void fun(immutable A a) {}              // 2
+
+struct B
+{
+    int a[];
+    @implicit this(ref A) shared {}             // ctor1
+    @implicit this(ref A) const {}              // ctor2
+    @implicit this(ref shared A) shared {}      // ctor3
+    @implicit this(ref immutable A) const {}    // ctor4
+}
+
+void gun(shared A) {}    // 3
+void gun(const A) {}     // 4
+
+void main()
+{
+    A a;
+    fun(a);             // calls 1 - conversion to const match; standard copying
+
+    const A a1;
+    fun(a1);             // calls 1 - direct match; standard copying
+
+    shared A a2;
+    fun(a2);             // calls 2 - conversion with copy ctor match; call copy ctor on parameter
+
+    B b;
+    fun(b);              // error: both gun functions match, because copy
+                         // ctor matches both ctor1 and ctor2
+
+    const B b1;
+    fun(b1)               // calls 4 - direct match; standard copying
+
+    immutable B b2;
+    fun(b2);              // calls 4 - conversion to const match; call ctor4
+
+    shared B b3;
+    fun(b3);              // calls 4 - direct match; call ctor3
+}
+```
+
+#### Interaction with `alias this`
+
+There are situations in which a struct defines both an `alias this` and a copy constructor, and
+for which assignments to variables of the struct type may lead to ambiguities:
+
+```d
+struct A
+{
+    int a;
+    immutable(A) fun()
+    {
+        return immutable A(7);
+    }
+
+    alias fun this;
+
+    @implicit this(ref A another) immutable {}
+}
+
+struct B
+{
+    int a;
+    A fun()
+    {
+        return A(7);
+    }
+
+    alias fun this;
+
+    @implicit this(ref B another) immutable {}
+
+}
+
+void main()
+{
+    immutable A ia;
+    A a = ia;            // 1 - calls copy constructor
+
+    B bc;
+    B b = bc;            // 2 - b is evaluated to B.fun
```

Comment:
AFAICT this is a declaration of b, not an "evalution". Did you mean to declare `b` of type `A`?

Reply:
Yes.